_posts/2016-10-26-jwt-artsy-journey.md

```diff
@@ -0,0 +1,80 @@
+---
+layout: post
+title: "JSON Web Tokens: Artsy's Journey"
+date: 2016-10-26 11:03
+comments: true
+author: ashkan
+categories: [Authentication, JWT, API]
+---
+
+At Artsy we currently have ~1730 client applications hitting our API and requesting authentication. When a user successfully authenticates through one of these clients, we want to embed basic user/application data in the resulting token rather than have to look up a session ID in the database on each request. For that we want to use JWT.
+
+JWT (JSON Web Token) is a self-contained, secure and standard way of transmitting data between applications and clients as JSON objects.
+<!-- more -->
+JWT has three separate sections which are separated by `.`: header, payload and signature.
+
+`<header>.<payload>.<signature>`
+
+Once decoded:
+
+- The header section contains JSON defining the type of token and how to treat its contents.
+
+- The payload section includes _claims_ and data provided by token issuer about the user/application, things like token expiration time, user id and etc.
+
+- Signature section is used to verify the sender of the JWT and make sure token wasn't modified along the way.
+
+![JWT example](/images/2016-10-26-JWT-Artsy-Journey/jwt-example.png)
+
+<a href="https://jwt.io/introduction/" target="_blank">JWT</a> and <a href="https://tools.ietf.org/html/rfc7519" target="_blank">RFC 7519</a> provide more details about each section of JWT and how encoding, decoding and verification work.
+
+
+# Where we were
+
+After a successful login, our API would generate a custom JSON object including some basic information about the user and application, then encrypt it using a single server-side secret.
+For every authenticated request we would have to decrypt the token, make sure user and application are valid and application still has access to our API. We are already _stateless_ since we don't store tokens in our database.
+However, client applications had to make an API call to get any information about the authenticated user.
+
+
+# Where we are going
+We want to keep our current auth flow which is already stateless and mainly replace our in-house generated access token with a more standard JWT. What that gives us is:
```

Comment:
> What that gives us is

None of the following items in the list would carry on from this sentence grammatically. Perhaps? 

> This let's us:



> Make Client applications who are able...
> Possibly include different...

> Follow a well-defined standard...